gpt_index/dataset_generation/base.py

```diff
@@ -0,0 +1,112 @@
+"""Dataset generation from documents"""
+from __future__ import annotations
+
+from typing import List, Optional
+import re
+
+from gpt_index import (
+    Document,
+    GPTListIndex,
+    QuestionAnswerPrompt,
+    ServiceContext,
+    SimpleDirectoryReader,
+    LLMPredictor,
+)
+
+from langchain.chat_models import ChatOpenAI
+from llama_index.langchain_helpers.text_splitter import TokenTextSplitter
+
+DEFAULT_QUESTION_GENERATION_PROMPT = """"Context information is below.\n"
+"\n---------------------\n{context_str}\n---------------------\n"
+"Given the context information and not prior knowledge.\n"
+"generate only questions based on the below query.\n"
+"{query_str}\n"
+"""
+
+
+class DatasetGenerator:
+    """Generate dataset (question/ question-answer pairs) \
+    based on the given documents.
+
+    NOTE: this is a beta feature, subject to change!
+
+    Args:
+        data_folder: Path to documents folder,
+        model_name: "gpt-3.5-turbo" or "gpt-4",
+        num_questions_per_chunk: number of question to be \
+        generated per chunk. Each document is chunked of size 512 words.
+        text_question_template: Question generation template.
+    """
+
+    def __init__(
+        self,
+        data_folder: Optional[str],
+        model_name: str = "gpt-3.5-turbo",
+        num_questions_per_chunk: int = 10,
+        text_question_template: Optional[QuestionAnswerPrompt] = None,
+    ) -> None:
+        """Init params."""
+        self.documents = SimpleDirectoryReader(data_folder).load_data()
+        self.model_name = model_name
+        self.num_questions_per_chunk = num_questions_per_chunk
+        self.text_question_template = text_question_template or QuestionAnswerPrompt(
+            DEFAULT_QUESTION_GENERATION_PROMPT
+        )
+        self.document_chunks = self.create_document_chunks()
+
+    def create_document_chunks(self) -> List[List[str]]:
+        """
+        Creates chunks for each document.
+        """
+        text_splitter = TokenTextSplitter(chunk_size=1000, chunk_overlap=100)
+
+        document_chunks = [
+            text_splitter.split_text(document.text) for document in self.documents
+        ]
+
+        return document_chunks
+
+    def generate_questions(self) -> List[List[str]]:
+        """
+        Generates questions for each document.
+        """
+
+        def document_question_generator(chunks: List[str]) -> List[str]:
```

Comment:
i wouldn't define this as a nested function. Define this on the class `_document_question_generator`